File: fragments/solana_program_username/solana_username_interface.ts
import {
  AccountRole,
  addDecoderSizePrefix,
  addEncoderSizePrefix,
  Address,
  appendTransactionMessageInstruction,
  assertAccountExists,
  Decoder,
  Encoder,
  fetchEncodedAccount,
  getAddressDecoder,
  getAddressEncoder,
  getArrayDecoder,
  getProgramDerivedAddress,
  getStructDecoder,
  getStructEncoder,
  getU32Decoder,
  getU32Encoder,
  getU64Decoder,
  getUtf8Decoder,
  getUtf8Encoder,
  KeyPairSigner,
  offsetDecoder,
} from "@solana/kit";
import { getInstructionDiscriminator, getPda, skipAnchorDiscriminator } from "../solana_program/solana_program_utils";
import { SYSTEM_PROGRAM_ADDRESS } from "@solana-program/system";
import { Buffer } from "node:buffer";
import { initRpcClient } from "../solana_rpc/solana_rpc_utils";
import {
  createBaseTxWithFeePayerAndLifetime,
  signAndSendTransaction,
} from "../solana_transaction/solana_transaction_utils";

export const initializeUsername = async (
  keypair: KeyPairSigner,
  programAddress: Address,
  username: Username["value"],
) => {
  const discriminator = getInstructionDiscriminator("initialize_username", "username");
  const payer = keypair.address;
  const usernamePda = await getPda(payer, programAddress, "user_account");
  const baseTx = await createBaseTxWithFeePayerAndLifetime(payer);

  // Avoid the readonly error due to type return of 'ReadonlyUint8Array'
  const encodedUsername = Buffer.from(usernameEncoder.encode({ value: username }));

  const tx = appendTransactionMessageInstruction({
    programAddress,
    data: Buffer.concat([discriminator, encodedUsername]),
    accounts: [
      { address: payer, role: AccountRole.WRITABLE_SIGNER },
      { address: usernamePda, role: AccountRole.WRITABLE },
      { address: SYSTEM_PROGRAM_ADDRESS, role: AccountRole.READONLY },
    ],
  }, baseTx);

  const sig = await signAndSendTransaction(tx, keypair.keyPair);

  return sig;
};

export const getUsernameAccount = async (keypair: KeyPairSigner, programAddress: Address) => {
  const client = initRpcClient();
  const usernamePda = await getPda(keypair.address, programAddress, "user_account");
  const account = await fetchEncodedAccount(client, usernamePda, {
    commitment: "confirmed",
    abortSignal: AbortSignal.timeout(5000),
  });

  if (!account.exists) {
    throw new Error(`Account ${usernamePda} does not exist`);
  }

  assertAccountExists(account);

  const decoded = usernameAccountDecoder.decode(account.data);

  return decoded;
};

export const updateUsername = async (keypair: KeyPairSigner, programAddress: Address, username: Username["value"]) => {
  const usernameAccount = await getUsernameAccount(keypair, programAddress);
  const changeCount = usernameAccount.change_count;
  const discriminator = getInstructionDiscriminator("update_username", "username");
  const payer = keypair.address;
  const usernamePda = await getPda(payer, programAddress, "user_account");
  const baseTx = await createBaseTxWithFeePayerAndLifetime(payer);
  const usernameRecordPda = await getUsernameRecordAccountPda(payer, programAddress, changeCount);

  // Avoid the readonly error due to type return of 'ReadonlyUint8Array'
  const encodedUsername = Buffer.from(usernameEncoder.encode({ value: username }));

  const tx = appendTransactionMessageInstruction({
    programAddress,
    data: Buffer.concat([discriminator, encodedUsername]),
    accounts: [
      { address: payer, role: AccountRole.WRITABLE_SIGNER },
      { address: usernamePda, role: AccountRole.WRITABLE },
      { address: usernameRecordPda, role: AccountRole.WRITABLE },
      { address: SYSTEM_PROGRAM_ADDRESS, role: AccountRole.READONLY },
    ],
  }, baseTx);

  const sig = await signAndSendTransaction(tx, keypair.keyPair);

  return sig;
};

interface Username {
  value: string;
}

export const getUsernameRecordAccount = async (
  keypair: KeyPairSigner,
  programAddress: Address,
  changeCount: bigint,
) => {
  const client = initRpcClient();
  const pda = await getUsernameRecordAccountPda(keypair.address, programAddress, changeCount);
  const account = await fetchEncodedAccount(client, pda, {
    commitment: "confirmed",
    abortSignal: AbortSignal.timeout(5000),
  });

  if (!account.exists) {
    throw new Error(`Account ${pda} does not exist`);
  }

  assertAccountExists(account);

  const decoded = usernameRecordAccountDecoder.decode(account.data);

  return decoded;
};

const getUsernameRecordAccountPda = async (userAddress: Address, programAddress: Address, changeCount: bigint) => {
  const changeCountSeed = Buffer.alloc(8);
  changeCountSeed.writeBigInt64LE(changeCount);
  const [pda] = await getProgramDerivedAddress({
    programAddress,
    seeds: [Buffer.from("username_record"), getAddressEncoder().encode(userAddress), changeCountSeed],
  });
  return pda;
};

const usernameEncoder: Encoder<{ value: string }> = getStructEncoder([
  ["value", addEncoderSizePrefix(getUtf8Encoder(), getU32Encoder())],
]);

const usernameDecoder = getStructDecoder([
  ["value", addDecoderSizePrefix(getUtf8Decoder(), getU32Decoder())],
]);

const usernameAccountDecoder: Decoder<{
  authority: Address;
  username: Username;
  change_count: bigint;
  username_recent_history: Username[];
}> = offsetDecoder(
  getStructDecoder([
    ["authority", getAddressDecoder()],
    ["username", usernameDecoder],
    ["change_count", getU64Decoder()],
    ["username_recent_history", getArrayDecoder(usernameDecoder, { size: getU32Decoder() })],
  ]),
  skipAnchorDiscriminator,
);

const usernameRecordAccountDecoder: Decoder<{
  authority: Address;
  old_username: Username;
  change_index: bigint;
}> = offsetDecoder(
  getStructDecoder([
    ["authority", getAddressDecoder()],
    ["old_username", usernameDecoder],
    ["change_index", getU64Decoder()],
  ]),
  skipAnchorDiscriminator,
);

File: fragments/solana_program_username/solana_username_interface.py
from typing import TypedDict
from solders.keypair import Keypair
from solders.pubkey import Pubkey
from solders.signature import Signature
from solders.instruction import Instruction, AccountMeta
from solana.constants import SYSTEM_PROGRAM_ID
from construct import Struct, Int32ul, Int64ul, PascalString, Bytes, PrefixedArray
from fragments.solana_program import get_instruction_discriminator, get_program_derived_address
from fragments.solana_rpc import init_rpc_client
from fragments.solana_transaction import create_tx_with_fee_payer_and_lifetime


class Username(TypedDict):
    value: str


class UsernameAccount(TypedDict):
    authority: Pubkey
    username: Username
    change_count: int
    username_recent_history: list[Username]


class UsernameRecordAccount(TypedDict):
    authority: Pubkey
    old_username: Username
    change_index: int


username_schema = Struct("value" / PascalString(Int32ul, encoding="utf-8"))

username_account_schema = Struct(
    "authority" / Bytes(32),
    "username" / username_schema,
    "change_count" / Int64ul,
    "username_recent_history" / PrefixedArray(Int32ul, username_schema),
)

username_record_account_schema = Struct(
    "authority" / Bytes(32),
    "old_username" / username_schema,
    "change_index" / Int64ul,
)


async def initialise_username(user_keypair: Keypair, program_address: Pubkey, username: str) -> Signature:
    discriminator = get_instruction_discriminator("initialize_username", "username")
    username_pda = get_program_derived_address(user_keypair.pubkey(), program_address, "user_account")
    client = init_rpc_client()
    encoded_username = Int32ul.build(len(username)) + username.encode("utf-8")
    instruction = Instruction(
        program_id=program_address,
        data=discriminator + encoded_username,
        accounts=[
            AccountMeta(pubkey=user_keypair.pubkey(), is_signer=True, is_writable=True),
            AccountMeta(pubkey=username_pda, is_signer=False, is_writable=True),
            AccountMeta(pubkey=SYSTEM_PROGRAM_ID, is_signer=False, is_writable=False),
        ],
    )
    tx = await create_tx_with_fee_payer_and_lifetime(user_keypair, instruction)
    response = await client.send_transaction(tx)
    return response.value


async def get_username_account(user_keypair: Keypair, program_address: Pubkey) -> UsernameAccount:
    client = init_rpc_client()
    pda = get_program_derived_address(user_keypair.pubkey(), program_address, "user_account")
    response = await client.get_account_info(pda)
    account_info = response.value

    if account_info is None:
        raise ValueError(f"Account {pda} does not exist")

    raw_bytes = bytes(account_info.data)[8:]  # removes the discriminator from the account data
    parsed = username_account_schema.parse(raw_bytes)

    authority = Pubkey.from_bytes(parsed.authority)
    username_history = parsed.get("username_recent_history", [])

    return UsernameAccount(
        authority=authority,
        username=parsed.username,
        change_count=parsed.change_count,
        username_recent_history=username_history,
    )


def get_username_record_pda(user_address: Pubkey, program_address: Pubkey, change_index: int) -> Pubkey:
    pda, _ = Pubkey.find_program_address(
        seeds=[b"username_record", bytes(user_address), change_index.to_bytes(8, "little")],
        program_id=program_address,
    )
    return pda


async def update_username(user_keypair: Keypair, program_address: Pubkey, username: str) -> Signature:
    username_account = await get_username_account(user_keypair, program_address)
    change_count = username_account["change_count"]
    discriminator = get_instruction_discriminator("update_username", "username")
    username_account_pda = get_program_derived_address(user_keypair.pubkey(), program_address, "user_account")
    username_record_account_pda = get_username_record_pda(user_keypair.pubkey(), program_address, change_count)
    encoded_username = Int32ul.build(len(username)) + username.encode("utf-8")
    client = init_rpc_client()
    instruction = Instruction(
        program_id=program_address,
        data=discriminator + encoded_username,
        accounts=[
            AccountMeta(pubkey=user_keypair.pubkey(), is_signer=True, is_writable=True),
            AccountMeta(pubkey=username_account_pda, is_signer=False, is_writable=True),
            AccountMeta(pubkey=username_record_account_pda, is_signer=False, is_writable=True),
            AccountMeta(pubkey=SYSTEM_PROGRAM_ID, is_signer=False, is_writable=False),
        ],
    )
    tx = await create_tx_with_fee_payer_and_lifetime(user_keypair, instruction)
    response = await client.send_transaction(tx)
    return response.value


async def get_username_record_account(
    user_address: Pubkey, program_address: Pubkey, change_index: int
) -> UsernameRecordAccount:
    client = init_rpc_client()
    pda = get_username_record_pda(user_address, program_address, change_index)
    response = await client.get_account_info(pda)
    account_info = response.value

    if account_info is None:
        raise ValueError(f"Account {pda} does not exist")

    raw_bytes = bytes(account_info.data)[8:]  # removes the discriminator from the account data
    parsed = username_record_account_schema.parse(raw_bytes)

    return UsernameRecordAccount(
        authority=Pubkey.from_bytes(parsed.authority),
        old_username=parsed.old_username,
        change_index=parsed.change_index,
    )
